dango/account-margin/src/query.rs

```diff
@@ -1,16 +1,86 @@
 use {
+    anyhow::bail,
     dango_auth::NEXT_SEQUENCE,
-    dango_types::account::single::QueryMsg,
-    grug::{ImmutableCtx, Json, JsonSerExt, StdResult, Storage},
+    dango_lending::DEBTS,
+    dango_oracle::OracleQueries,
+    dango_types::{
+        account::margin::{CollateralPower, HealthResponse, QueryMsg},
+        config::AppConfig,
+    },
+    grug::{
+        Addr, BorshDeExt, Coins, Denom, ImmutableCtx, Inner, IsZero, Json, JsonSerExt, NumberConst,
+        QuerierWrapper, StdResult, Storage, Udec128,
+    },
+    std::collections::BTreeMap,
 };
 
 #[cfg_attr(not(feature = "library"), grug::export)]
-pub fn query(ctx: ImmutableCtx, msg: QueryMsg) -> StdResult<Json> {
+pub fn query(ctx: ImmutableCtx, msg: QueryMsg) -> anyhow::Result<Json> {
     match msg {
         QueryMsg::Sequence {} => query_sequence(ctx.storage)?.to_json_value(),
+        QueryMsg::Health {} => query_health(ctx)?.to_json_value(),
     }
+    .map_err(Into::into)
 }
 
 fn query_sequence(storage: &dyn Storage) -> StdResult<u32> {
     NEXT_SEQUENCE.current(storage)
 }
+
+/// Calculates the health of a margin account.
+pub fn calculate_account_health(
+    querier: &QuerierWrapper,
+    margin_account: Addr,
+    debts: Coins,
+    collateral_powers: BTreeMap<Denom, CollateralPower>,
+) -> anyhow::Result<HealthResponse> {
+    let app_cfg: AppConfig = querier.query_app_config()?;
+
+    // Calculate the total value of the debts.
+    let total_debt_value = debts.into_iter().try_fold(Udec128::ZERO, |acc, coin| {
+        let price = querier.query_price(app_cfg.addresses.oracle, &coin.denom)?;
+        Ok::<_, anyhow::Error>(acc + price.value_of_unit_amount(coin.amount)?)
+    })?;
```

Comment:
I think this code is just so much worse than just using a `for`-loop.

Especially the `Ok::<_, anyhow::Error>` part

Reply:
That's what I thought, but @Rhaki had strong opinions that this was better, so I changed it.